base/broadcast.jl

```diff
@@ -859,4 +859,229 @@ macro __dot__(x)
+#                FusionCall(
+#                  *,
+#                  ( FusionArg{2}(),
+#                    FusionArg{3}() )), )), )),
+#        (:a, :b, :c))
+## and then the final expansion becomes:
+# d = broadcast(d', a, b, c)
+
+struct Fusion{N, vararg#=::Bool=#, T}
+    f::T
+    # Debugging Metadata:
+    # names::NTuple{N, Symbol}
+    # source::LineNumberNode
+    function Fusion{N, vararg}(f) where {N, vararg}
+        return new{N, vararg::Bool, typeof(f)}(f)
+    end
+end
+
+struct FusionArg{N}
+end
+
+struct FusionConstant{T}
+    c::T
+    function FusionConstant(c) where {}
+        return new{typeof(c)}(c)
+    end
+end
+
+struct FusionCall{F, Args<:Tuple}
+    f::F
+    args::Args
+    function FusionCall(f, args::Tuple) where {}
+        return new{typeof(f), typeof(args)}(f, args)
+    end
+end
+
+struct FusionApply{N, F, Args<:NTuple{N, Any}}
+    f::F
+    args::Args
+    function FusionApply(f, args::NTuple{N, Any}) where {N}
+        return new{N, typeof(f), typeof(args)}(f, args)
+    end
+end
+
+function kw_to_vec(kws::Vector{Any})
+    kwargs = Vector{Any}(2 * length(kws))
+    for i in 1:2:length(kws)
+        kw = kws[i]::Tuple{Any, Any}
+        kwargs[i] = getfield(kw, 1)
+        kwargs[i + 1] = getfield(kw, 2)
+    end
+    return kwargs
+end
+
+struct FusionKWCall{F, Args<:Tuple}
+    f::F
+    args::Args
+    kwargs::Vector{Any}
+    function FusionKWCall(f, args::Tuple; kwargs...) where {}
+        return new{typeof(f), typeof(args)}(f, args, kw_to_vec(kwargs))
+    end
+end
+
+struct FusionKWApply{F, Args<:Tuple}
+    f::F
+    args::Args
+    kwargs::Vector{Any}
+    function FusionKWApply(f, args::Tuple; kwargs...)  where {}
+        return new{typeof(f), typeof(args)}(f, args, kw_to_vec(kwargs))
+    end
+end
+
+function tuplehead(t::Tuple, N::Val)
+    return ntuple(i -> t[i], N)
+end
+@generated function tupletail(t::NTuple{M, Any}, ::Val{N}) where {N, M}
+    # alternative, non-generated versions,
+    # enable when inference is improved:
+    #tupletail(t, Nreq) = ntuple(i -> t[i + Nreq], length(t) - Nreq)
+    #tupletail(t, Nreq) = t[(Nreq + 1):end]
+    args = Any[ :(getfield(t, $i)) for i in (N + 1):M ]
+    tpl = Expr(:tuple)
+    tpl.args = args
+    return tpl
+end
+
+@inline (f::Fusion{N, false})(args::Vararg{Any, N}) where {N} = f.f(args...)
+function (f::Fusion{Nreq, true})(args::Vararg{Any, M}) where {Nreq, M}
+    M >= Nreq || throw(MethodError(f, args))
+    fargs = tuplehead(args, Val(Nreq))
+    vararg = tupletail(args, Val(Nreq))
+    return f.f((fargs..., vararg)...)
+end
+@inline (f::FusionArg{N})(args...) where {N} = args[N]
+@inline (f::FusionConstant)(args...) = f.c
+@inline (f::FusionCall)(args...) = f.f(map(a -> a(args...), f.args)...)
+# TODO: calling _apply on map _apply is not handled by inference
+# for now, we unroll some cases and generate others, to help it out
+#@inline (f::FusionApply)(args...) = Core._apply(f.f, map(a -> a(args...), f.args)...)
```

Comment:
equivalently, I think this is also `f.f(flatten(map(a -> a(args...), f.args)...)...)`
where `flatten(args...) = Core._apply(tuple, args...)`